sled-agent/src/artifact_store.rs

```diff
@@ -441,7 +451,38 @@ async fn ledger_manager(
             Ledger::<ArtifactConfig>::new(&log, ledger_paths.clone()).await
         {
             if new_config.generation > ledger.data().generation {
-                // New config generation; update the ledger.
+                // New config generation. First check that it's not asking
+                // us to delete any artifact that is part of the current zone
+                // configuration.
+                if let Some(services) = services {
+                    let mut difference = ledger
+                        .data()
+                        .artifacts
+                        .difference(&new_config.artifacts)
+                        .copied()
+                        .peekable();
+                    if difference.peek().is_some() {
+                        let in_use = services
+                            .omicron_zones_list()
+                            .await
+                            .zones
+                            .into_iter()
+                            .filter_map(|zone| {
+                                zone.image_source.artifact_hash()
+                            })
+                            .collect::<BTreeSet<_>>();
+                        for sha256 in difference {
```

Comment:
I think this is fine but I had a couple thoughts reading it:

1. I think this assumes something else guarantees that the artifact store ledger always contain artifacts for every zone in the zone ledger, right? If somehow we had a zone that wasn't in the ledger, we wouldn't catch it being missing from `new_config` here.
2. In light of that - do we need to compute `difference` at all? Or could we check `in_use` against `new_config` directly?

Reply:
Yes, this would only catch artifacts removed from one generation of the config to the next. It does seem to make more sense to reject any configurations that are missing in-use artifacts, probably with a clearer error message.